src/v/archival/ntp_archiver_service.cc

```diff
@@ -1929,10 +2181,35 @@ ss::future<> ntp_archiver::apply_spillover() {
     }
 }
 
+bool ntp_archiver::stm_retention_needed() const {
+    auto arch_so = manifest().get_archive_start_offset();
+    // Return true if there is no archive
+    return arch_so == model::offset{};
+}
+
 ss::future<> ntp_archiver::apply_retention() {
     if (!may_begin_uploads()) {
         co_return;
     }
+    auto arch_so = manifest().get_archive_start_offset();
+    auto stm_so = manifest().get_start_offset();
+    if (arch_so != model::offset{} && arch_so != stm_so) {
+        // We shouldn't do retention in the part of the log controlled by
+        // the archival STM if archive region is not empty. It's unlikely for
+        // STM retention and archive retention to work together. Most of the
+        // time either STM retention will be happening without spillover or
+        // archive retention will be happening alongside spillover from the STM.
+        // This is just a safety check that prevents situation when the method
+        // is called for log with not empty archive. In this case the retention
+        // will remove too much data from the STM region of the log.
+        vlog(
+          _rtclog.warn,
+          "Archive start offset {} is not equal to STM start offset {}, "
+          "skipping STM retention",
+          arch_so,
+          stm_so);
+        co_return;
```

Comment:
As an extra layer of defence against applying retention on the in-memory manifest, can we make `retention_calculator::next_start_offset` return `std::nullopt` if the archive is initialised?